Fix server error when `blowfish_secret` is not exactly 32 bytes long

MauricioFauth opened the pull request:
- [x] Back-port commit c5611531c03a5c9dfe13510e7fc5b901daa77039
- [x] Update documentation about the `blowfish_secret` setting
- [x] Update the setup page to reflect the `blowfish_secret` changes

Fixes https://github.com/phpmyadmin/phpmyadmin/issues/17369.

MauricioFauth committed:
Remove `trigger_error` calls from `HomeController`

Signed-off-by: Maurício Meneghini Fauth <mauricio@fauth.dev>
(cherry picked from commit c5611531c03a5c9dfe13510e7fc5b901daa77039)
Signed-off-by: Maurício Meneghini Fauth <mauricio@fauth.dev>

williamdes reviewed (approved):
Was there a technical reason for not using trigger error, I am not sure I remember

MauricioFauth commented:
> Was there a technical reason for not using trigger error, I am not sure I remember 

The trigger_error is converted to a error exception.
With this way, the error is still shown, but without the need for a error/exception handler.

I'll also update the documentation to reflect the changes and update the setup page to provide better and more secure secrets.

MauricioFauth committed:
Update setup page to generate better blowfish_secret keys

Instead of generating a printable string, it generates a binary string
and converts it to an hexadecimal string.

Signed-off-by: Maurício Meneghini Fauth <mauricio@fauth.dev>

MauricioFauth commented on libraries/classes/Setup/ConfigGenerator.php:
```diff
@@ -94,6 +100,12 @@ public static function getConfigFile(ConfigFile $cf)
+        if ($var_name === 'blowfish_secret') {
+            $secret = self::getBlowfishSecretKey($var_value);
+
+            return sprintf('$cfg[\'blowfish_secret\'] = \sodium_hex2bin(\'%s\');%s', sodium_bin2hex($secret), $crlf);
```
I think that having a function call ([`sodium_hex2bin`](https://www.php.net/manual/en/function.sodium-hex2bin.php)) in the config file is a good trade-off for having better keys.

williamdes replied:
we could use binary strings but in the format `0xfffffabcd0123` and it would not need a call to an extension. And that would keep our config files future proof and have no "logic"

MauricioFauth replied:
Could you provide an example? Since `0xfffffabcd0123` is treated as a number.
```
php > var_dump(hex2bin('48656c6c6f20776f726c6421'));
php shell code:1:
string(12) "Hello world!"
php > var_dump(0x48656c6c6f20776f726c6421);
php shell code:1:
double(2.2405534230754E+28)
php > var_dump((string) 0x48656c6c6f20776f726c6421);
php shell code:1:
string(19) "2.2405534230754E+28"
```

MauricioFauth replied:
> it would not need a call to an extension

It's a bundled extension since PHP 7.2, so it's not an exotic extension. That's why I think it's safe to add this function call. And `sodium_hex2bin` is constant-time, which is important for encryption keys to avoid attacks, while `hex2bin` is not.

williamdes replied:
> It's a bundled extension since PHP 7.2, so it's not an exotic extension.

That's a good point. Maybe at some point we should remove the need for the compat lib ?
https://github.com/phpmyadmin/phpmyadmin/blob/b9d1abb1013a4719fe8b7d2ea4cca6b2fb59374f/composer.json#L53

> And sodium_hex2bin is constant-time, which is important for encryption keys to avoid attacks, while hex2bin is not.

Can you explain a bit more how this brings more security ?
I mean most of our users will not have `sodium_hex2bin` in their config files so how does this work our for actual users vs new ones ?
Should people use `sodium_hex2bin` in their config to be more secure ?
Do we add a message to ask people to change ?

MauricioFauth replied:
> That's a good point. Maybe at some point we should remove the need for the compat lib ?

Yes, I think we can remove the `paragonie/sodium_compat` package in the future.

> Can you explain a bit more how this brings more security ?

If you converted the string to hexadecimal, then when converting back to binary, the `sodium_hex2bin` function will prevent the key being leaked by a timing attack.

> I mean most of our users will not have sodium_hex2bin in their config files so how does this work our for actual users vs new ones ?
> Should people use sodium_hex2bin in their config to be more secure ?
> Do we add a message to ask people to change ?

Since the `blowfish_secret` was already 32 chars long, there is no need to change as it's already a valid key. However, when using a binary string, it's possible to generate better keys because there is more possibilities. But since it's not printable, it should be converted to hexadecimal or base64 before pasting it in the config file.

williamdes replied:
I just had an idea, could we let the config with the hex and in our code call the sodium function?
Like that the config would still remain logic less

MauricioFauth replied:
And what about the users that already have a valid key? We'll have to support a 64-chars long hexadecimal string and a 32-bytes long string. The user may have this key saved in another place, for example in a file, they will have to convert to hexadecimal first?

williamdes replied:
For valid keys we can detect that it's not a 64 hex and assume it's a 32 one
For users with files they can do what they want in their config file as before nothing changes
From a config file point of view nothing changes. Only allowing 64 hex keys
In our code we handle each format and use or not the sodium lib

Would that work?

MauricioFauth replied:
We could support both formats. But I think this is more confusing.

> In our code we handle each format and use or not the sodium lib

We use the sodium lib regardless. For the 64 hex case, the function call will only move from the config file to the code that handles it.

williamdes replied:
My fear is that some day it will crash in our faces and we will have to ask users to update the config because we can not do anything about it. If you think it is best to have it like you proposed that will be okay for me and we will see what is better in the end after some years :)

MauricioFauth replied:
Oh, I see. For example, the sodium_hex2bin being deprecated or we changing the method to a new one and having to asking the users to update the config.

williamdes replied:
> Oh, I see. For example, the sodium_hex2bin being deprecated or we changing the method to a new one and having to asking the users to update the config.

Yup, exactly. For now the config file traveled 10+ years without anything breaking and no updates needed ^^

MauricioFauth replied:
Except when I broke it when I added a requirement for a 32 bytes string. I changed the code and now it accepts longer strings, but will only use the first 32 bytes.
The sodium_hex2bin function is not required in the config file, only a 32 bytes long string. So the current values will keep working seamlessly. My intention in suggesting using this function is to help users having better keys by using bytes, instead of only characters.
The Sodium is a bundled extension, which means that it's part of the language. So a possible deprecation will be hopefully well planned. That way, we'll have time to work on a solution if we need to.

williamdes replied:
Okay, let's do that. Let's say to users they can do one or another while updating keys

MauricioFauth commented on libraries/classes/Config/ServerConfigChecks.php:
```diff
@@ -345,55 +348,21 @@ protected function performConfigChecksCookieAuthUsed(
-            return;
-        }
-
-        $blowfishWarnings = [];
-        // check length
-        if (strlen($blowfishSecret) < 32) {
-            // too short key
-            $blowfishWarnings[] = __('Key is too short, it should have at least 32 characters.');
-        }
-
-        // check used characters
-        $hasDigits = (bool) preg_match('/\d/', $blowfishSecret);
-        $hasChars = (bool) preg_match('/\S/', $blowfishSecret);
-        $hasNonword = (bool) preg_match('/\W/', $blowfishSecret);
-        if (! $hasDigits || ! $hasChars || ! $hasNonword) {
-            $blowfishWarnings[] = Sanitize::sanitizeMessage(
-                __(
-                    'Key should contain letters, numbers [em]and[/em] special characters.'
-                )
-            );
-        }
```
This was removed because it generates a new key not only when the key is not set, but when it does not have the correct length as well. And since it's a binary string, does not makes sense to check for used characters in the string.

MauricioFauth committed:
Update the docs to reflect the blowfish_secret changes

Signed-off-by: Maurício Meneghini Fauth <mauricio@fauth.dev>

williamdes commented on doc/config.rst:
```diff
@@ -1896,25 +1896,42 @@ Cookie authentication options
-    prompted for this passphrase.
+    The "cookie" auth\_type uses the Sodium extension to encrypt the cookies. If you are using the "cookie" auth\_type,
+    enter here a generated string of random bytes to be used as an encryption key. It will be used internally by the
+    Sodium extension: you won’t be prompted for this encryption key.
```
```suggestion
    Sodium extension: you won't be prompted for this encryption key.
```

MauricioFauth committed:
Document how to generate a string of random bytes

Signed-off-by: Maurício Meneghini Fauth <mauricio@fauth.dev>

MauricioFauth commented:
Hi, @ibennetch. Could you please review the wording of the documentation changes?

MauricioFauth committed:
Allow longer cookie encryption keys to be used

Signed-off-by: Maurício Meneghini Fauth <mauricio@fauth.dev>

MauricioFauth committed:
Add a note that sodium_bin2hex is not required for blowfish_secret

Signed-off-by: Maurício Meneghini Fauth <mauricio@fauth.dev>

williamdes reviewed (approved):
🚢

williamdes commented on libraries/classes/Controllers/HomeController.php:
```diff
@@ -292,39 +296,43 @@ private function checkRequirements(): void
+                    'message' => __(
+                        'The configuration file needs a valid key for cookie encryption.'
+                        . ' A temporary key was automatically generated for you.'
+                        . ' Please refer to the [doc@cfg_blowfish_secret]documentation[/doc].'
```
I am not on my workstation but was wondering if the link target did have a citation for users to find the FAQ entry you added
